File: src/store.ts
// --- Store convention ---
// The context store IS the protocol. Every agent is a directory on disk with a known layout:
//   CONTEXT.md  — working memory (mutable, private)
//   PROFILE.md  — public address card (replaces SOUL.md: "soul" implied private identity,
//                 but this file is shared with peers — "profile" is honest about its visibility)
//   inbox/      — append-only message queue from other agents
//   outbox/     — signed envelopes waiting to be delivered
//   shared/     — files explicitly published to peers
//   history/    — conversation logs
//   knowledge/  — reference docs
//   .keys/      — Ed25519 + age keypairs (gitignored)
//   .mesh.json  — config, peer list, keyring
// No database, no daemon required. `ls` is your status command.

import { readFile, writeFile, mkdir, readdir, appendFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { existsSync } from "node:fs";
import {
  generateKeys, hasKeys, signMessage, signAndEncrypt, verifyMessage, decryptMessage,
  deserializeSignedMessage, serializeSignedMessage, wrapExternalMessage,
  fingerprint, type SignedMessage, type KeyringEntry,
} from "./crypto.js";

export interface MeshConfig {
  id: string;
  name: string;
  created: string;
  publicKey?: string;
  encryptionKey?: string;
  peers: PeerConfig[];
  keyring: KeyringEntry[];
  trustedKeys?: string[]; // legacy v0.1 flat list — auto-migrated to keyring on first read
  autoTrust?: boolean; // workspace mode: auto-trust all imported keys
}

export interface PeerConfig {
  id: string;
  name: string;
  url: string;
  access: "read" | "readwrite";
  mountPath?: string;
}

const STORE_DIRS = ["history", "knowledge", "inbox", "outbox", "shared", ".peers"];

/** Validate agent/peer names: alphanumeric + hyphens + underscores + dots, 1-64 chars.
 *  Rejects path traversal (../, /, \) and rsync glob chars (*, ?, [). */
export function validateName(name: string, label = "Name"): string {
  if (!name || name.length < 1 || name.length > 64) {
    throw new Error(`${label} must be 1-64 characters`);
  }
  if (!/^[a-zA-Z0-9][a-zA-Z0-9._-]*$/.test(name)) {
    throw new Error(`${label} must start with alphanumeric and contain only a-z, 0-9, -, _, .`);
  }
  if (name.includes("..") || name.includes("/") || name.includes("\\")) {
    throw new Error(`${label} contains invalid path characters`);
  }
  return name;
}

/** Resolve a keyring entry by name, name:fingerprint, or bare fingerprint prefix.
 *  Throws if ambiguous (multiple matches) or not found. */
export function resolveKeyring(keyring: KeyringEntry[], query: string): KeyringEntry {
  let name: string;
  let fpPrefix: string | undefined;

  if (query.includes(":")) {
    // name:FINGERPRINT format — split on LAST colon group that looks like hex
    const colonIdx = query.lastIndexOf(":");
    const maybeFp = query.slice(colonIdx + 1);
    if (/^[0-9a-fA-F]{4,16}$/.test(maybeFp)) {
      name = query.slice(0, colonIdx);
      fpPrefix = maybeFp.toUpperCase();
    } else {
      name = query;
    }
  } else {
    name = query;
  }

  // Match by name (or address prefix)
  let matches = keyring.filter(
    (k) => k.name === name || k.address.startsWith(`${name}@`)
  );

  // If no name match, try bare fingerprint prefix
  if (matches.length === 0 && /^[0-9a-fA-F]{4,16}$/.test(query)) {
    const upper = query.toUpperCase();
    matches = keyring.filter(
      (k) => k.fingerprint.replace(/:/g, "").startsWith(upper)
    );
  }

  // Filter by fingerprint prefix if provided
  if (fpPrefix && matches.length > 1) {
    matches = matches.filter(
      (k) => k.fingerprint.replace(/:/g, "").startsWith(fpPrefix!)
    );
  }

  if (matches.length === 0) {
    throw new Error(`Key not found: "${query}". Run: openfuse key list`);
  }
  if (matches.length > 1) {
    const options = matches.map(
      (k) => `  ${k.name}:${k.fingerprint.replace(/:/g, "").slice(0, 8)}  ${k.address}`
    ).join("\n");
    throw new Error(
      `Multiple keys match "${query}". Disambiguate with fingerprint:\n${options}`
    );
  }
  return matches[0];
}

export class ContextStore {
  readonly root: string;

  constructor(root: string) {
    this.root = resolve(root);
  }

  get configPath() {
    return join(this.root, ".mesh.json");
  }

  async exists(): Promise<boolean> {
    return existsSync(this.configPath);
  }

  async init(name: string, id: string): Promise<void> {
    await mkdir(this.root, { recursive: true });
    for (const dir of STORE_DIRS) {
      await mkdir(join(this.root, dir), { recursive: true });
    }

    // Copy templates
    const templatesDir = new URL("../templates/", import.meta.url).pathname;
    for (const file of ["CONTEXT.md", "PROFILE.md"]) {
      const templatePath = join(templatesDir, file);
      const destPath = join(this.root, file);
      if (!existsSync(destPath)) {
        const content = await readFile(templatePath, "utf-8");
        await writeFile(destPath, content);
      }
    }

    const keys = await generateKeys(this.root);

    const config: MeshConfig = {
      id,
      name,
      created: new Date().toISOString(),
      publicKey: keys.publicKey,
      encryptionKey: keys.encryptionKey,
      peers: [],
      keyring: [],
    };
    await this.writeConfig(config);
  }

  // Shared workspace: multiple agents mount the same directory.
  // CHARTER.md = system prompt (purpose, rules). CONTEXT.md = shared working memory.
  // tasks/ for coordination, messages/{agent}/ for DMs, _broadcast/ for all-hands.
  async initWorkspace(name: string, id: string): Promise<void> {
    await mkdir(this.root, { recursive: true });
    for (const dir of ["tasks", "messages", "_broadcast", "shared", "history"]) {
      await mkdir(join(this.root, dir), { recursive: true });
    }

    const templatesDir = new URL("../templates/", import.meta.url).pathname;
    for (const file of ["CHARTER.md", "CONTEXT.md"]) {
      const templatePath = join(templatesDir, file);
      const destPath = join(this.root, file);
      if (!existsSync(destPath)) {
        const content = await readFile(templatePath, "utf-8");
        await writeFile(destPath, content);
      }
    }

    // Workspaces auto-trust: all imported keys are trusted by default.
    // Safe because workspaces are private — you control who joins.
    const config: MeshConfig = {
      id,
      name,
      created: new Date().toISOString(),
      peers: [],
      keyring: [],
      autoTrust: true,
    };
    await this.writeConfig(config);
  }

  async readConfig(): Promise<MeshConfig> {
    const raw = await readFile(this.configPath, "utf-8");
    const config = JSON.parse(raw) as MeshConfig;

    // Migrate legacy trustedKeys → keyring (v0.1 stored bare public keys in a flat array;
    // v0.2+ uses a GPG-style keyring with trust levels, fingerprints, and encryption keys)
    if (config.trustedKeys && config.trustedKeys.length > 0) {
      if (!config.keyring) config.keyring = [];
      for (const key of config.trustedKeys) {
        const k = key.trim();
        if (!k || config.keyring.some((e) => e.signingKey === k)) continue;
        config.keyring.push({
          name: `migrated-${k.slice(0, 8)}`,
          address: "",
          signingKey: k,
          fingerprint: fingerprint(k),
          trusted: true,
          added: new Date().toISOString(),
        });
      }
      delete config.trustedKeys;
      await this.writeConfig(config);
    }

    if (!config.keyring) config.keyring = [];
    return config;
  }

  async writeConfig(config: MeshConfig): Promise<void> {
    await writeFile(this.configPath, JSON.stringify(config, null, 2) + "\n");
  }

  async readContext(): Promise<string> {
    return readFile(join(this.root, "CONTEXT.md"), "utf-8");
  }

  async writeContext(content: string): Promise<void> {
    await writeFile(join(this.root, "CONTEXT.md"), content);
  }

  // --- Context compaction ---
  // Agents mark sections as [DONE] when work is complete. `openfuse compact`
  // moves done sections to history/YYYY-MM-DD.md, keeping CONTEXT.md lean.
  // Sections are delimited by markdown headers (## or ###).

  async compactContext(): Promise<{ moved: number; kept: number }> {
    const content = await this.readContext();
    const lines = content.split("\n");
    const kept: string[] = [];
    const done: string[] = [];
    let current: string[] = [];
    let currentDone = false;

    const flush = () => {
      if (current.length > 0) {
        (currentDone ? done : kept).push(current.join("\n"));
        current = [];
        currentDone = false;
      }
    };

    for (const line of lines) {
      if (/^#{1,3}\s/.test(line)) {
        flush();
        currentDone = /\[DONE\]/i.test(line);
      }
      current.push(line);
    }
    flush();

    if (done.length === 0) return { moved: 0, kept: kept.length };

    // Write kept sections back to CONTEXT.md
    await this.writeContext(
      kept.join("\n\n") || "# Context\n\n*Working memory — what's happening right now.*\n"
    );

    // Append done sections to history/YYYY-MM-DD.md
    const historyDir = join(this.root, "history");
    await mkdir(historyDir, { recursive: true });
    const dateStr = new Date().toISOString().split("T")[0];
    const historyFile = join(historyDir, `${dateStr}.md`);
    const header = existsSync(historyFile) ? "\n---\n\n" : `# Context History — ${dateStr}\n\n`;
    await appendFile(historyFile, header + done.join("\n\n") + "\n");

    return { moved: done.length, kept: kept.length };
  }

  async readProfile(): Promise<string> {
    return readFile(join(this.root, "PROFILE.md"), "utf-8");
  }

  async writeProfile(content: string): Promise<void> {
    await writeFile(join(this.root, "PROFILE.md"), content);
  }

  // --- Inbox ---

  async sendInbox(peerId: string, message: string): Promise<string> {
    const config = await this.readConfig();

    // Resolve recipient from keyring — supports name, name:fingerprint, or bare fingerprint.
    // Throws if ambiguous or not found.
    const entry = resolveKeyring(config.keyring, peerId);

    let signed: SignedMessage;
    if (entry.encryptionKey) {
      signed = await signAndEncrypt(this.root, config.name, message, entry.encryptionKey);
    } else {
      signed = await signMessage(this.root, config.name, message);
    }

    const shortFp = entry.fingerprint.replace(/:/g, "").slice(0, 8);
    const recipientDir = `${peerId}-${shortFp}`;
    const outboxDir = join(this.root, "outbox", recipientDir);
    await mkdir(outboxDir, { recursive: true });

    const timestamp = new Date().toISOString().replace(/[:.]/g, "-");
    const filename = `${timestamp}_from-${config.name}.json`;
    await writeFile(join(outboxDir, filename), serializeSignedMessage(signed));
    return `${recipientDir}/${filename}`;
  }

  async readInbox(): Promise<Array<{
    file: string;
    content: string;
    wrappedContent: string;
    from: string;
    time: string;
    verified: boolean;
    encrypted: boolean;
  }>> {
    const inboxDir = join(this.root, "inbox");
    if (!existsSync(inboxDir)) return [];

    const config = await this.readConfig();
    const files = await readdir(inboxDir);
    const messages = [];

    for (const file of files.filter((f) => f.endsWith(".json") || f.endsWith(".md"))) {
      const raw = await readFile(join(inboxDir, file), "utf-8");

      const signed = deserializeSignedMessage(raw);
      if (signed) {
        const sigValid = verifyMessage(signed);
        // Identity binding: verify BOTH that the key is trusted AND that the claimed
        // sender name matches the name we associated with that key in our keyring.
        // Without this, a trusted agent could forge the "from" field and impersonate
        // someone else while still showing [VERIFIED].
        const keyMatchesName = (k: typeof config.keyring[0]) =>
          k.signingKey.trim() === signed.publicKey.trim() &&
          (k.name === signed.from || k.address.startsWith(`${signed.from}@`));
        const trusted = config.autoTrust
          ? config.keyring.some(keyMatchesName)
          : config.keyring.some((k) => k.trusted && keyMatchesName(k));
        const verified = sigValid && trusted;

        let content: string;
        if (signed.encrypted) {
          try {
            content = await decryptMessage(this.root, signed);
          } catch {
            content = "[encrypted — cannot decrypt]";
          }
        } else {
          content = signed.message;
        }

        messages.push({
          file,
          content,
          wrappedContent: wrapExternalMessage(signed, verified),
          from: signed.from,
          time: signed.timestamp,
          verified,
          encrypted: !!signed.encrypted,
        });
      } else {
        const parts = file.replace(/\.(md|json)$/, "").split("_");
        const from = parts.slice(1).join("_");
        messages.push({
          file,
          content: raw,
          wrappedContent: wrapExternalMessage(
            { from, timestamp: parts[0], message: raw, signature: "", publicKey: "" },
            false,
          ),
          from,
          time: parts[0],
          verified: false,
          encrypted: false,
        });
      }
    }

    return messages.sort((a, b) => a.time.localeCompare(b.time));
  }

  // --- Shared files ---

  async listShared(): Promise<string[]> {
    const sharedDir = join(this.root, "shared");
    if (!existsSync(sharedDir)) return [];
    return readdir(sharedDir);
  }

  async share(filename: string, content: string): Promise<void> {
    // Path traversal defense: basename extraction + ".." rejection.
    // Critical because MCP tools pass user-supplied filenames directly.
    const base = filename.split("/").pop()!.split("\\").pop()!;
    if (!base || base === "." || base === ".." || base.includes("..")) {
      throw new Error(`Invalid filename: ${filename}`);
    }
    const sharedDir = join(this.root, "shared");
    await mkdir(sharedDir, { recursive: true });
    await writeFile(join(sharedDir, base), content);
  }

  // --- Status ---

  async status(): Promise<{
    id: string;
    name: string;
    peers: number;
    inboxCount: number;
    sharedCount: number;
  }> {
    const config = await this.readConfig();
    const inbox = await this.readInbox();
    const shared = await this.listShared();
    return {
      id: config.id,
      name: config.name,
      peers: config.peers.length,
      inboxCount: inbox.length,
      sharedCount: shared.length,
    };
  }
}

File: rust/src/store.rs
use std::fs;
use std::path::{Path, PathBuf};
use serde::{Deserialize, Serialize};
use anyhow::Result;

use crate::crypto;

const STORE_DIRS: &[&str] = &["history", "knowledge", "inbox", "outbox", "shared"];

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct MeshConfig {
    pub id: String,
    pub name: String,
    pub created: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub public_key: Option<String>,
    pub peers: Vec<PeerConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trusted_keys: Option<Vec<String>>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PeerConfig {
    pub id: String,
    pub name: String,
    pub url: String,
    pub access: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mount_path: Option<String>,
}

pub struct StatusInfo {
    pub id: String,
    pub name: String,
    pub peers: usize,
    pub inbox_count: usize,
    pub shared_count: usize,
}

pub struct InboxMessage {
    pub file: String,
    pub content: String,
    pub wrapped_content: String,
    pub from: String,
    pub time: String,
    pub verified: bool,
}

pub struct ContextStore {
    root: PathBuf,
}

impl ContextStore {
    pub fn new(root: &Path) -> Self {
        Self { root: root.to_path_buf() }
    }

    pub fn root(&self) -> &Path {
        &self.root
    }

    pub fn config_path(&self) -> PathBuf {
        self.root.join(".mesh.json")
    }

    pub fn exists(&self) -> bool {
        self.config_path().exists()
    }

    pub fn init(&self, name: &str, id: &str) -> Result<()> {
        fs::create_dir_all(&self.root)?;
        for dir in STORE_DIRS {
            fs::create_dir_all(self.root.join(dir))?;
        }

        // Write template files if they don't exist
        let context_path = self.root.join("CONTEXT.md");
        if !context_path.exists() {
            fs::write(&context_path, "# Context\n\n*Working memory — what's happening right now.*\n")?;
        }

        let soul_path = self.root.join("SOUL.md");
        if !soul_path.exists() {
            fs::write(&soul_path, format!("# Soul\n\n*Agent identity and rules.*\n\n**Name:** {}\n**ID:** {}\n", name, id))?;
        }

        // Generate signing keys
        let public_key = crypto::generate_keys(&self.root)?;

        // Write config
        let config = MeshConfig {
            id: id.to_string(),
            name: name.to_string(),
            created: chrono::Utc::now().to_rfc3339(),
            public_key: Some(public_key),
            peers: vec![],
            trusted_keys: Some(vec![]),
        };
        self.write_config(&config)?;
        Ok(())
    }

    pub fn read_config(&self) -> Result<MeshConfig> {
        let raw = fs::read_to_string(self.config_path())?;
        Ok(serde_json::from_str(&raw)?)
    }

    pub fn write_config(&self, config: &MeshConfig) -> Result<()> {
        let json = serde_json::to_string_pretty(config)?;
        fs::write(self.config_path(), format!("{}\n", json))?;
        Ok(())
    }

    pub fn read_context(&self) -> Result<String> {
        Ok(fs::read_to_string(self.root.join("CONTEXT.md"))?)
    }

    pub fn write_context(&self, content: &str) -> Result<()> {
        fs::write(self.root.join("CONTEXT.md"), content)?;
        Ok(())
    }

    pub fn read_soul(&self) -> Result<String> {
        Ok(fs::read_to_string(self.root.join("SOUL.md"))?)
    }

    pub fn write_soul(&self, content: &str) -> Result<()> {
        fs::write(self.root.join("SOUL.md"), content)?;
        Ok(())
    }

    pub fn send_inbox(&self, peer_id: &str, message: &str, from: &str) -> Result<()> {
        let signed = crypto::sign_message(&self.root, from, message)?;
        let serialized = serde_json::to_string_pretty(&signed)?;
        let timestamp = chrono::Utc::now().to_rfc3339().replace([':', '.'], "-");
        let filename = format!("{}_{}.json", timestamp, peer_id);
        fs::write(self.root.join("inbox").join(&filename), &serialized)?;
        fs::write(self.root.join("outbox").join(&filename), &serialized)?;
        Ok(())
    }

    pub fn read_inbox(&self) -> Result<Vec<InboxMessage>> {
        let inbox_dir = self.root.join("inbox");
        if !inbox_dir.exists() {
            return Ok(vec![]);
        }

        let config = self.read_config()?;
        let trusted_keys: Vec<String> = config.trusted_keys
            .unwrap_or_default()
            .iter()
            .map(|k| k.trim().to_string())
            .collect();

        let mut messages = vec![];

        for entry in fs::read_dir(&inbox_dir)? {
            let entry = entry?;
            let path = entry.path();
            let fname = path.file_name().and_then(|n| n.to_str()).unwrap_or("").to_string();

            if !fname.ends_with(".json") && !fname.ends_with(".md") {
                continue;
            }

            let raw = fs::read_to_string(&path)?;

            if let Ok(signed) = serde_json::from_str::<crypto::SignedMessage>(&raw) {
                if signed.from.is_empty() || signed.message.is_empty() {
                    continue;
                }
                let sig_valid = crypto::verify_message(&signed);
                let trusted = trusted_keys.iter().any(|k| k == signed.public_key.trim());
                let verified = sig_valid && trusted;
                let wrapped = crypto::wrap_external_message(&signed, verified);
                messages.push(InboxMessage {
                    file: fname,
                    content: signed.message.clone(),
                    wrapped_content: wrapped,
                    from: signed.from.clone(),
                    time: signed.timestamp.clone(),
                    verified,
                });
            } else {
                // Unsigned fallback
                let stem = fname.trim_end_matches(".json").trim_end_matches(".md");
                let parts: Vec<&str> = stem.splitn(2, '_').collect();
                let from = if parts.len() > 1 { parts[1].to_string() } else { "unknown".to_string() };
                let time = parts[0].to_string();
                let wrapped = format!(
                    "<external_message from=\"{}\" verified=\"false\" time=\"{}\" status=\"UNVERIFIED\">\n{}\n</external_message>",
                    from, time, raw
                );
                messages.push(InboxMessage {
                    file: fname,
                    content: raw,
                    wrapped_content: wrapped,
                    from,
                    time,
                    verified: false,
                });
            }
        }

        messages.sort_by(|a, b| a.time.cmp(&b.time));
        Ok(messages)
    }

    pub fn list_shared(&self) -> Result<Vec<String>> {
        let shared_dir = self.root.join("shared");
        if !shared_dir.exists() {
            return Ok(vec![]);
        }
        let files = fs::read_dir(shared_dir)?
            .filter_map(|e| e.ok())
            .filter_map(|e| e.file_name().into_string().ok())
            .collect();
        Ok(files)
    }

    pub fn share(&self, filename: &str, content: &str) -> Result<()> {
        let shared_dir = self.root.join("shared");
        fs::create_dir_all(&shared_dir)?;
        fs::write(shared_dir.join(filename), content)?;
        Ok(())
    }

    pub fn status(&self) -> Result<StatusInfo> {
        let config = self.read_config()?;
        let inbox = self.read_inbox()?;
        let shared = self.list_shared()?;
        Ok(StatusInfo {
            id: config.id,
            name: config.name,
            peers: config.peers.len(),
            inbox_count: inbox.len(),
            shared_count: shared.len(),
        })
    }
}
